File: Java/src/_0073_Set_Matrix_Zeroes/Solution.java
package _0073_Set_Matrix_Zeroes;

import _0000_study._common.Output;

import java.util.HashSet;

/**
 * Date: 2020/9/18 16:18
 * Content:
 */
public class Solution {

    public void setZeroes(int[][] matrix) {

        int m = matrix.length;
        if (m == 0) return;
        int n = matrix[0].length;
        // 第一步：记录数值为 0 的行号 和 列号
        HashSet<Integer> rowSet = new HashSet<>();
        HashSet<Integer> columnSet = new HashSet<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == 0) {
                    rowSet.add(i);
                    columnSet.add(j);
                }
            }
        }

        // 第二步：将 0 所在的行列数据清零
        for (Integer integer : rowSet) {
            for (int j = 0; j < n; j++)
                matrix[integer][j] = 0;
        }
        for (Integer integer : columnSet) {
            for (int i = 0; i < m; i++)
                matrix[i][integer] = 0;
        }
    }

    public static void main(String[] args) {

        int[][] matrix = {
                {0, 1, 2, 0},
                {3, 4, 5, 2},
                {1, 3, 1, 5}
        };
        (new Solution()).setZeroes(matrix);
        Output.OutputBasicArray2D(matrix);
    }
}

File: Python/_0073_Set_Matrix_Zeroes/Solution.py
# coding=utf-8
# Date: 2021/5/26 16:55
from typing import List


class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix);
        m = len(matrix[0])
        r0 = False;
        c0 = False
        for i in range(m):
            if matrix[0][i] == 0:
                r0 = True
                break
        for i in range(n):
            if matrix[i][0] == 0:
                c0 = True
                break

        for i in range(1, n):
            for j in range(m):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    break
        for i in range(1, m):
            for j in range(n):
                if matrix[j][i] == 0:
                    matrix[0][i] = 0
                    break

        for i in range(1, n):
            if matrix[i][0] == 0:
                for j in range(1, m):
                    matrix[i][j] = 0
        for i in range(1, m):
            if matrix[0][i] == 0:
                for j in range(n):
                    matrix[j][i] = 0
        if r0:
            for i in range(m):
                matrix[0][i] = 0
        if c0:
            for i in range(n):
                matrix[i][0] = 0


if __name__ == "__main__":
    matrix = [[1, 1, 1], [1, 0, 1], [1, 1, 1]]
    Solution().setZeroes(matrix)
    print(matrix)  # [[1, 0, 1], [0, 0, 0], [1, 0, 1]]
